### PR TITLE
make migration backup folder a sibling of the project root

### PR DESCRIPTION
**Pull request by jonsequitur**
This changes the migration backup folder to be a peer of the root project being migrated. Each project being migrated will be placed under the `backup` folder using the source folder's name.  

It fixes issue #4757.

### CODE REVIEW

File: src/Microsoft.DotNet.ProjectJsonMigration/Microsoft.DotNet.Internal.ProjectModel/Utilities/PathUtility.cs [hunk 1]
```diff
@@ -54,9 +54,15 @@ private static string EnsureTrailingCharacter(string path, char trailingCharacte
         public static void EnsureParentDirectory(string filePath)
         {
             string directory = Path.GetDirectoryName(filePath);
-            if (!Directory.Exists(directory))
+
+            EnsureDirectory(directory);
+        }
+        
+        public static void EnsureDirectory(string directoryPath)
```
Comment (livarcocc) [inline src/Microsoft.DotNet.ProjectJsonMigration/Microsoft.DotNet.Internal.ProjectModel/Utilities/PathUtility.cs]: Create a issue to track it then. So that it does not get lost.

File: src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            if (workspaceDirectory == null)
+            {
+                throw new ArgumentNullException(nameof(workspaceDirectory));
+            }
+
+            globalJson = new FileInfo(Path.Combine(
+                           workspaceDirectory.FullName,
+                           "global.json"));
+
+            projectDirectory = new DirectoryInfo(projectDirectory.FullName.EnsureTrailingSlash());
```
Comment (livarcocc) [inline src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs]: Why is this needed?

File: src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        public DirectoryInfo ProjectBackupDirectory { get; }
+
+        public DirectoryInfo RootBackupDirectory { get; }
+
+        public IEnumerable<FileInfo> FilesToMove { get; }
+
+        public void PerformBackup()
+        {
```
Comment (krwq) [inline src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs]: will this preserve directory structure in the backup folder?

Comment (jonsequitur) [inline src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs]: Yes.

File: src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+            foreach (var file in FilesToMove)
+            {
+                file.MoveTo(
+                    Path.Combine(
+                        ProjectBackupDirectory.FullName, file.Name));
+            }
+        }
+    }
+}
```
Comment (krwq) [inline src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs]: nit: line ending

File: src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs (hunk 2 above)
Comment (jonsequitur) [inline src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs]: Which part?

Comment (livarcocc) [inline src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs]: The EnsureTrailingSlash part.

File: src/Microsoft.DotNet.ProjectJsonMigration/Microsoft.DotNet.Internal.ProjectModel/Utilities/PathUtility.cs (hunk 1 above)
Comment (jonsequitur) [inline src/Microsoft.DotNet.ProjectJsonMigration/Microsoft.DotNet.Internal.ProjectModel/Utilities/PathUtility.cs]: I'll just include it in the current PR. It's a small enough change.

File: src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        public IEnumerable<FileInfo> FilesToMove { get; }
+
+        public void PerformBackup()
+        {
+            if (globalJson.Exists)
+            {
+                PathUtility.EnsureDirectory(RootBackupDirectory.FullName);
+
+                globalJson.MoveTo(Path.Combine(
+                    ProjectBackupDirectory.Parent.FullName,
```
Comment (livarcocc) [inline src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs]: Should this be RootBackupDirectory? It is the one that you are ensuring exists.

File: src/Microsoft.DotNet.TestFramework/TestDirectory.cs [hunk 6]
```diff
@@ -27,9 +27,19 @@ internal TestDirectory(string path)
 
         private static void EnsureExistsAndEmpty(string path)
         {
-            if (Directory.Exists(path))
+            var testDirectory = new DirectoryInfo(path);
+
+            var migrationBackupDirectory = new DirectoryInfo(
+                System.IO.Path.Combine(testDirectory.Parent.FullName, "backup"));
```
Comment (livarcocc) [inline src/Microsoft.DotNet.TestFramework/TestDirectory.cs]: put System.IO in the using namespaces?

File: src/Microsoft.DotNet.TestFramework/TestDirectory.cs [hunk 7]
```diff
@@ -27,9 +27,19 @@ internal TestDirectory(string path)
 
         private static void EnsureExistsAndEmpty(string path)
```
Comment (livarcocc) [inline src/Microsoft.DotNet.TestFramework/TestDirectory.cs]: Should you indicate in the name that you will also handle backup folders?

Comment (jonsequitur) [inline src/Microsoft.DotNet.TestFramework/TestDirectory.cs]: Good suggestion.

File: src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs (hunk 2 above)
Comment (jonsequitur) [inline src/Microsoft.DotNet.ProjectJsonMigration/MigrationBackupPlan.cs]: Ah. I'm normalizing the input since there's no guarantee that the caller created it with a trailing slash. `DirectoryInfo` doesn't enforce it and its presence or absence will be visible in the `FullName` property.

**Comment by jonsequitur** [comment]
@dotnet-bot Test Ubuntu16.04 x64 Debug Build​

**Comment by jonsequitur** [comment]
@dotnet-bot Test Debian8.2 x64 Debug Build

**Comment by livarcocc** [comment]
@srivatsn

File: test/dotnet-migrate.Tests/GivenThatIWantToMigrateTestApps.cs [hunk 8]
```diff
@@ -654,17 +657,19 @@ private void BuildProjectJson(string projectDirectory)
...
         private void MigrateProject(params string[] migrateArgs)
         {
-            var result =
-                MigrateCommand.Run(migrateArgs);
-
-            result.Should().Be(0);
+            new TestCommand("dotnet")
+                    .WithForwardingToConsole()
+                    .Execute($"migrate {string.Join(" ", migrateArgs)}")
+                    .Should()
```
Comment (piotrpMSFT) [inline test/dotnet-migrate.Tests/GivenThatIWantToMigrateTestApps.cs]: I like to keep each assertion clause on one line. Putting `.Should` and `.Pass` on two lines reduces readability as additional clauses are added e.g.:

```
.Should()
.Pass()
.And
.HaveFiles()
```

vs. 

```
.Should().Pass()
.And.HaveFiles()
```

Thoughts?

File: test/Microsoft.DotNet.ProjectJsonMigration.Tests/MigrationBackupPlanTests.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
Comment (piotrpMSFT) [inline test/Microsoft.DotNet.ProjectJsonMigration.Tests/MigrationBackupPlanTests.cs]: I'd like to see a test which uses a global.json file to define the solution root. I believe these tests assume the root is the parent directory of the starting project.

It would also be useful to have a test which does a restore of some root directory. E.g.

```
/core
  /github
    /piotrp
      /cli
        global.json
```

In the above, I should be able to run migrate from any directory above `global.json` and have migration pick up the root directory as the directory that contains global.json

Finally, I'm not sure if global.json's `src` property supported relative paths to a parent directory. If it did, it would be good to understand how we handle those since the `backup` directory's layout cannot handle `..`'s afaik.

File: test/dotnet-migrate.Tests/GivenThatIWantToMigrateTestApps.cs (hunk 8 above)
Comment (jonsequitur) [inline test/dotnet-migrate.Tests/GivenThatIWantToMigrateTestApps.cs]: If you want to avoid long lines, `Should` often needs its own line, e.g. when followed by an assertion method with a longer name, so I tend to go more vertical for the sake of consistent formatting. It's just a matter of taste though.
